services/mailer/mail_issue.go

```diff
@@ -112,6 +109,20 @@ func mailIssueCommentToParticipants(issue *models.Issue, doer *models.User, acti
 	return nil
 }
 
+func addUniqueUsers(visited map[string]bool, current []*models.User, list []*models.User) []*models.User {
```

Comment:
if we colect only the IDs (int) first and call models.GetUsersByIDs(recipients_ids) once we dont need this can save recources too since we only deal with int until we need the UserList ...

Reply:
@6543 I started with that, but then I needed to change a lot of functions in `models`, tests, etc.. In fact, I renamed that branch "fix-8952-the-hard-way" and dropped it half way.

Reply:
@6543 I've ended doing what you've suggested. Now look at the complexity of the PR. 😉